File: templates/python/discovery/local_discovery.py
"""
LocalDiscovery — in-process registry for development & testing.

Swap this for:
  HttpDiscovery    → production REST registry
  GossipDiscovery  → P2P gossip network
  OnChainDiscovery → ERC-8004 Ethereum registry
"""

from interfaces.iagent_discovery import IAgentDiscovery, DiscoveryEntry
from datetime import datetime, timezone
from typing import Dict, List


class LocalDiscovery(IAgentDiscovery):
    _instance: "LocalDiscovery | None" = None
    _registry: Dict[str, DiscoveryEntry]

    def __init__(self):
        self._registry = {}

    @classmethod
    def get_instance(cls) -> "LocalDiscovery":
        """Singleton — share one registry per process."""
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance

    async def register(self, entry: DiscoveryEntry) -> None:
        entry.registered_at = datetime.now(timezone.utc).isoformat()
        self._registry[entry.agent_id] = entry
        print(f"[LocalDiscovery] Registered: {entry.agent_id} ({', '.join(entry.capabilities)})")

    async def unregister(self, agent_id: str) -> None:
        self._registry.pop(agent_id, None)
        print(f"[LocalDiscovery] Unregistered: {agent_id}")

    async def query(self, capability: str) -> List[DiscoveryEntry]:
        return [
            e for e in self._registry.values()
            if capability in e.capabilities and e.health.status != "unhealthy"
        ]

    async def list_all(self) -> List[DiscoveryEntry]:
        return list(self._registry.values())

    async def heartbeat(self, agent_id: str) -> None:
        entry = self._registry.get(agent_id)
        if entry:
            entry.health.last_heartbeat = datetime.now(timezone.utc).isoformat()
            entry.health.status = "healthy"

File: templates/typescript/discovery/LocalDiscovery.ts
import { IAgentDiscovery, DiscoveryEntry } from '../interfaces/IAgentDiscovery';

/**
 * LocalDiscovery — in-process registry for development & testing.
 *
 * Swap this for:
 *   - HttpDiscovery     → production REST registry
 *   - GossipDiscovery   → P2P gossip network
 *   - OnChainDiscovery  → ERC-8004 Ethereum registry
 */
export class LocalDiscovery implements IAgentDiscovery {
  private static instance: LocalDiscovery;
  private registry = new Map<string, DiscoveryEntry>();

  private constructor() {}

  /** Singleton — share one registry per process */
  static getInstance(): LocalDiscovery {
    if (!LocalDiscovery.instance) {
      LocalDiscovery.instance = new LocalDiscovery();
    }
    return LocalDiscovery.instance;
  }

  async register(entry: DiscoveryEntry): Promise<void> {
    this.registry.set(entry.agentId, { ...entry, registeredAt: new Date().toISOString() });
    console.log(`[LocalDiscovery] Registered: ${entry.agentId} (${entry.capabilities.join(', ')})`);
  }

  async unregister(agentId: string): Promise<void> {
    this.registry.delete(agentId);
    console.log(`[LocalDiscovery] Unregistered: ${agentId}`);
  }

  async query(capability: string): Promise<DiscoveryEntry[]> {
    return [...this.registry.values()].filter(e =>
      e.capabilities.includes(capability) && e.health.status !== 'unhealthy'
    );
  }

  async listAll(): Promise<DiscoveryEntry[]> {
    return [...this.registry.values()];
  }

  async heartbeat(agentId: string): Promise<void> {
    const entry = this.registry.get(agentId);
    if (entry) {
      entry.health.lastHeartbeat = new Date().toISOString();
      entry.health.status = 'healthy';
    }
  }
}
